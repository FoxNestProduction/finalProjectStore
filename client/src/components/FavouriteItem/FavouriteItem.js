import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import Typography from '@mui/material/Typography';
import { CardActions, CardContent, CardMedia, Rating, Button, Box, Card } from '@mui/material';
import AddBoxOutlinedIcon from '@mui/icons-material/AddBoxOutlined';
import ColorChips from '../Chip/Chip';
import { chipForFavourite } from '../Chip/styles';
import { stylesButton, styleCardFavourite, styleMediaFavourite, styleContentFavourite } from './styles';
import FavouriteIcon from '../FavouriteIcon/FavouriteIcon';
import { fixedEncodeURIComponent } from '../../utils/uriEncodeHelpers';
import { addProductToCart } from '../../redux/slices/cartSlice';

const FavouriteItem = ({ product }) => {
  const dispatch = useDispatch();

  const {
    name,
    itemNo,
    currentPrice,
    isTrending,
    rating,
    imageUrl,
    isSupreme,
    isHealthy,
    _id,
  } = product;

  const isFavourite = useSelector((state) => state.favourites.cardStates[_id]);

  const handleAddToCart = () => {
    dispatch(addProductToCart(_id));
  };

  return (
    <Card sx={styleCardFavourite}>
      <Link to={`/menu/${fixedEncodeURIComponent(name)}/${itemNo}`}>
        <Box sx={{ display: 'flex', justifyContent: 'space-between', maxWidth: '900px' }}>
          <CardMedia
            component="img"
            width="fit content"
            height="100%"
            image={imageUrl}
            sx={styleMediaFavourite}
          />
          <CardContent sx={styleContentFavourite}>
            <Box sx={{ display: 'flex', flexDirection: 'column', gap: '2px' }}>
              <Typography variant="h3" sx={{ p: 0, mb: 1 }}>{name}</Typography>
              <ColorChips
                isTrending={isTrending}
                isSupreme={isSupreme}
                isHealthy={isHealthy}
                customStyles={chipForFavourite}
              />
              <Rating name="half-rating" size="small" value={rating} readOnly sx={{ color: 'primary.main' }} />
            </Box>
            <Box>
              <Typography
                variant="h3"
              >
                $
                {currentPrice}
              </Typography>
            </Box>
          </CardContent>
        </Box>
      </Link>
      <CardActions sx={{ alignItems: 'flex-end', p: 0, minWidth: '130px' }}>
        <Button
          onClick={handleAddToCart}
          variant="outline"
          fontSize="medium"
          sx={stylesButton}
        >
          Add To Cart
          <AddBoxOutlinedIcon
            fontSize="small"
            sx={{ ml: 1 }}
          />
        </Button>
      </CardActions>
      <CardActions sx={{ position: 'absolute', top: '0', right: '0' }}>
        <FavouriteIcon id={_id} />
        {/* <FavouriteIcon product={product} /> */}
      </CardActions>
    </Card>
  );
};

FavouriteItem.propTypes = {
  product: PropTypes.object,
};
FavouriteItem.defaultProps = {
  product: {},
};

export default FavouriteItem;
