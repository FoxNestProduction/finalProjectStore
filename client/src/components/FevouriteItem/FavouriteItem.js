import React from 'react';
import { Link } from 'react-router-dom';
import { useSelector, shallowEqual } from 'react-redux';
import PropTypes from 'prop-types';
import Typography from '@mui/material/Typography';
import { CardActions, CardContent, CardMedia, Rating, Button, Box, Card } from '@mui/material';
import AddBoxOutlinedIcon from '@mui/icons-material/AddBoxOutlined';
import ColorChips from '../Chip/Chip';
import { chipForFavourite } from '../Chip/styles';
import { stylesButton, styleCardFavourite, styleMediaFavourite, styleContentFavourite } from './styles';
import FavouriteIcon from '../FavouriteIcon/FavouriteIcon';
import { fixedEncodeURIComponent } from '../../utils/uriEncodeHelpers';

const FavouriteItem = ({ favourite }) => {
  const products = useSelector((state) => state.products.products, shallowEqual);
  // const favourite = useSelector((state) => state.favourites.favourites);
  // const dish = useSelector((state) => state.favourites.favourites, shallowEqual);
  // eslint-disable-next-line no-underscore-dangle
  const dish = products.find((item) => (item._id) === favourite);

  const {
    name,
    currentPrice,
    isTranding,
    rating,
    imageUrl,
    isSupreme,
    isHealthy,
    _id,
  } = dish;

  return (
    <Card sx={styleCardFavourite}>
      {/* <Link to={`/menu/${fixedEncodeURIComponent(name)}`}> */}
      <Box sx={{ display: 'flex', justifyContent: 'space-between' }}>
        <CardMedia
          component="img"
          width="auto"
          height="100%"
          image={imageUrl}
          sx={styleMediaFavourite}
        />
        <CardContent sx={styleContentFavourite}>
          <Box sx={{ display: 'flex', flexDirection: 'column', gap: '2px' }}>
            <Typography variant="h3" sx={{ p: 0, mb: 1 }}>{name}</Typography>
            <ColorChips
              isTranding={isTranding}
              isSupreme={isSupreme}
              isHealthy={isHealthy}
              customStyles={chipForFavourite}
            />
            <Rating name="half-rating" size="small" value={rating} readOnly sx={{ color: 'primary.main' }} />
          </Box>
          <Box>
            <Typography
              variant="h3"
            >
              $
              {currentPrice}
            </Typography>
          </Box>
        </CardContent>
      </Box>
      {/* </Link> */}
      <CardActions sx={{ alignItems: 'flex-end', p: 0 }}>
        <Button
          variant="outline"
          fontSize="medium"
          sx={stylesButton}
        >
          Add To Cart
          <AddBoxOutlinedIcon
            fontSize="small"
            sx={{ ml: 1 }}
          />
        </Button>
      </CardActions>
      <CardActions sx={{ position: 'absolute', top: '0', right: '0' }}>
        <FavouriteIcon id={_id} />
      </CardActions>
    </Card>
  );
};

FavouriteItem.propTypes = {
  favourite: PropTypes.string.isRequired,
};

export default FavouriteItem;
