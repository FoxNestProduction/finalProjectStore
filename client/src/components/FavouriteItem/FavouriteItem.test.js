import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import FavouriteItem from './FavouriteItem';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
  Link: ({ to, children }) => <a href={to}>{children}</a>,
}));

describe('Snapshot test', () => {
  const mockDispatch = jest.fn();

  const product = {
    name: 'Product',
    itemNo: '123',
    currentPrice: 10.99,
    isTrending: true,
    rating: 4,
    imageUrl: 'image.jpg',
    isSupreme: false,
    isHealthy: true,
    _id: '123',
  };

  test('should FavouriteItem render', () => {
    const { asFragment } = render(<FavouriteItem product={product} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should FavouriteItem added to cart', () => {
    const useDispatchMock = jest.spyOn(require('react-redux'), 'useDispatch');
    useDispatchMock.mockReturnValue(mockDispatch);

    const { getByRole } = render(<FavouriteItem product={product} />);
    const activeButton = getByRole('button', { name: 'Add To Cart' });
    fireEvent.click(activeButton);
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'cart/addToCart',
      payload: {
        product: {
          _id: '123',
          currentPrice: 10.99,
          imageUrl: 'image.jpg',
          name: 'Product',
        },
        cartQuantity: 1,
      },
    });
  })
});
