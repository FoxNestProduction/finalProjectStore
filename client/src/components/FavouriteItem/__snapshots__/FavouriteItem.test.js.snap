// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test should FavouriteItem render 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-b839p-MuiPaper-root-MuiCard-root"
  >
    <a
      href="/menu/product/123"
    >
      <div
        class="MuiBox-root css-1kxi74m"
      >
        <img
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1xe5xw-MuiCardMedia-root"
          height="100%"
          src="image.jpg"
          width="fit content"
        />
        <div
          class="MuiCardContent-root css-7ozjol-MuiCardContent-root"
        >
          <div
            class="MuiBox-root css-3a3hw0"
          >
            <h3
              class="MuiTypography-root MuiTypography-h3 css-1mpegdo-MuiTypography-root"
            >
              Product
            </h3>
            <div
              class="MuiBox-root css-1ma6dsf"
            >
              <div
                class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-1m1nliy-MuiChip-root"
              >
                <span
                  class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                >
                  Healthy
                </span>
              </div>
              <div
                class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-lw0im9-MuiChip-root"
              >
                <span
                  class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                >
                  Trending
                </span>
              </div>
            </div>
            <span
              aria-label="4 Stars"
              class="MuiRating-root MuiRating-sizeSmall Mui-readOnly MuiRating-readOnly css-k076s9-MuiRating-root"
              role="img"
            >
              <span>
                <span
                  class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
              <span>
                <span
                  class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
              <span>
                <span
                  class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
              <span>
                <span
                  class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
              <span>
                <span
                  class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                    data-testid="StarBorderIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                    />
                  </svg>
                </span>
              </span>
            </span>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <h3
              class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
            >
              $10.99
            </h3>
          </div>
        </div>
      </div>
    </a>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-vgxr4e-MuiCardActions-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outline MuiButton-outlinePrimary MuiButton-sizeMedium MuiButton-outlineSizeMedium MuiButton-root MuiButton-outline MuiButton-outlinePrimary MuiButton-sizeMedium MuiButton-outlineSizeMedium css-1y82v5b-MuiButtonBase-root-MuiButton-root"
        font-size="medium"
        tabindex="0"
        type="button"
      >
        Add To Cart
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-2xhw3-MuiSvgIcon-root"
          data-testid="AddBoxOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14zm-8-2h2v-4h4v-2h-4V7h-2v4H7v2h4z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-mk70d3-MuiCardActions-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-u1vl96-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1gick0k-MuiSvgIcon-root"
          data-testid="FavoriteBorderOutlinedIcon"
          focusable="false"
          isactive="false"
          ishovered="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
